Splash my explorations server redirect

526avijitgupta opened the pull request:
Fixes #1533 

1. Server-side redirect based on current login status.
2. Update unit-tests to handle redirection (302)

seanlip commented on core/controllers/admin_test.py:
```diff
@@ -128,45 +128,49 @@ def test_change_rights(self):
-        self.testapp.get(feconf.GALLERY_URL).mustcontain(
-            '/moderator', no=['/admin'])
-        self.logout()
+        if self.testapp.get('/').status_int != 302:
```
This does not seem correct. A test should be checking expected behavior -- so, either a redirect happens, or it doesn't, and the test should know what is expected.

The user has not logged in at this point, so we do not expect a redirect, and the test should assert there is no redirect. When the user logs in, we expect a redirect, and the test should assert that there is one, etc.

526avijitgupta replied:
@seanlip I had a query:
In the current codebase, '/' does not have a separate page associated with it (it is directly pointed to gallery page now).

But now since checking for user login status has to be made for '/', so a redirect will have to be made, either to '/gallery' (if not logged in) or to '/my_explorations' (if logged in).  But it will always be a redirect.
Is this approach correct ?

seanlip replied:
Yes, that sounds good to me.

526avijitgupta replied:
@seanlip Sorry for being not very clear before, but your comment [here](https://github.com/oppia/oppia/pull/1547#discussion_r54686743)  and the approach above (apparatently correct), don't seem to go hand in hand. 

According to my approach, there will always be a redirect (even if the user is logged out), but your comment actually sounds like just the opposite of that.

seanlip replied:
Oh, sorry -- yes, that's my error. My main point was more that there should be no 'if' statement in the expected behavior of the tests, but you are right that I was inconsistent in my description of the behavior.

So, to be clear: I think that, if the user is not logged in, and visits '/', there should be a redirect to '/gallery'. If the user is logged in, and visits '/', there should be a redirect to '/my_explorations'.

seanlip commented on core/controllers/home_test.py:
```diff
@@ -26,10 +26,15 @@ class HomePageTest(test_utils.GenericTestBase):
-        response.mustcontain(
-            'Your personal tutor',
-            'Oppia - Gallery', 'About', 'Sign in', no=['Logout'])
+        if response.status_int != 302:
```
Ditto.

526avijitgupta replied:
Resolved similar to above

seanlip commented:
Hi @526avijitgupta -- I left a comment about the tests. Also, please note that there may be some existing references to "/"  in the frontend that should now be changed to "/gallery". Thanks!

526avijitgupta commented:
@seanlip Yes you are correct indeed. I'll fix the tests and submit another PR. Does that work ?

seanlip commented:
Yep, sounds good, thanks. Note that it's OK to make follow-on commits to this PR.

526avijitgupta commented:
If it's ok, then I'd rather make follow on-commits to this PR only.

seanlip commented:
Yes, that's fine -- just make additional commits to this branch, and they'll be reflected in the PR.

526avijitgupta commented:
@seanlip Update the unit tests according our discussion above. Please review.

seanlip commented on core/controllers/admin_test.py:
```diff
@@ -128,12 +128,12 @@ def test_change_rights(self):
-        # Navigate to any page. The role is not set.
-        self.testapp.get('/').mustcontain(no=['/moderator', '/admin'])
+        # There should be a redirect on navigating to /
+        self.assertEqual(self.testapp.get('/').status_int, 302)
```
Actually, with the new structure, you probably don't need these two lines here. I recommend adding another test altogether (in base_test.py) that checks that the redirection happens correctly in both the logged-out and logged-in cases, and that the destination pages are correct.

seanlip replied:
Oops, sorry, I take that back. I think you still want to add a new test as described, but the original line should instead be changed to

```
    self.testapp.get('/gallery').mustcontain(no=['/moderator', '/admin'])
```

The point of those lines was to check that the correct URLs show in the page depending on the user's rights.

526avijitgupta replied:
Replaced the original line with the one you just posted above.
Also, wrote a new test in base_test.py that checks for redirection in both logged in and logged out states.

seanlip commented on core/controllers/admin_test.py:
```diff
@@ -128,12 +128,12 @@ def test_change_rights(self):
+        # Log in as a superadmin. There should again be a redirect on /
         self.login('superadmin@example.com', is_super_admin=True)
-        self.testapp.get('/').mustcontain('/admin', no=['/moderator'])
+        self.assertEqual(self.testapp.get('/').status_int, 302)
```
Change this one to

    self.testapp.get('/gallery').mustcontain('/admin', no=['/moderator'])

526avijitgupta replied:
Done

seanlip commented on core/controllers/admin_test.py:
```diff
@@ -154,19 +154,22 @@ def test_change_rights(self):
 
         # Log in as a moderator.
         self.login(self.MODERATOR_EMAIL)
+        self.assertEqual(self.testapp.get('/').status_int, 302)
```
See above -- this whole test isn't really testing for correct redirection (that's not the point), so you don't need this line. Ditto below.

526avijitgupta replied:
Removed checking for redirection from both places

seanlip commented on core/controllers/home_test.py:
```diff
@@ -26,10 +26,7 @@ class HomePageTest(test_utils.GenericTestBase):
-        response.mustcontain(
-            'Your personal tutor',
-            'Oppia - Gallery', 'About', 'Sign in', no=['Logout'])
+        self.assertEqual(response.status_int, 302)
```
For this one, you do need to check that the destination location is correct, and then navigate to it (response.follow()? I'm not sure, you'll need to look up the webtest docs) and then check that it has the right terms. Don't just delete the original test.

526avijitgupta replied:
Used `response.follow()` method for the same, and kept the original test intact

seanlip commented:
Thanks @526avijitgupta! Made some comments.

526avijitgupta commented:
@seanlip Thanks for such a detailed review :+1: 

I've updated the tests based on your comments above. I think now it should be good to merge.

Also, the 3rd point in your comment [here](https://github.com/oppia/oppia/issues/1533#issuecomment-190518521), talks about a potential snag. 
Will there be a separate issue or PR for that, or should I simply continue making commits to this branch altogether for fixing that too ?

seanlip commented on core/controllers/home_test.py:
```diff
@@ -26,8 +26,8 @@ class HomePageTest(test_utils.GenericTestBase):
         response = self.testapp.get('/')
-        self.assertEqual(response.status_int, 200)
-        response.mustcontain(
+        self.assertEqual(response.status_int, 302)
```
Still need to assert correct response URL location. This is done elsewhere in other python tests where we do a 302; you might want to grep.

526avijitgupta replied:
Asserted the response location through [this](https://github.com/526avijitgupta/oppia/commit/af9a9544081eb1500d912dea24a7ee786296af5d) commit

seanlip commented:
Hi @526avijitgupta, the 3rd point of the comment should be addressed in this PR before it can be merged. In addition, please add the test described [here](https://github.com/oppia/oppia/pull/1547#discussion_r54863599).

Also, see previous comments in this thread: in particular, there may be some existing references to "/" in the frontend that should now be changed to "/gallery". This needs to be addressed too.

Finally, can you please respond to all comments following the protocol in step 5 [here](https://github.com/oppia/oppia/blob/develop/CONTRIBUTING.md#instructions-for-making-a-code-change)? This helps ensure nothing falls through the cracks.

Thanks!

526avijitgupta commented:
@seanlip Addressed the 3rd point of the comment by the commits - [1](https://github.com/526avijitgupta/oppia/commit/8f9de533ec4e7a3958669edeb9a661a325a2c868) and [2](https://github.com/oppia/oppia/commit/574a4c7ce15bd5b6d0b657ad833b8293bdbbf5e1)

Added the test (already mentioned above, in one of my previous comments).

Replaced reference(s) to '/' that needed to be changed to '/gallery' in [this](https://github.com/oppia/oppia/commit/63b81702e9aa5f8c27b5f13f0d982ba5c5bd3d88) commit.

526avijitgupta commented:
@seanlip Following instruction [#5](https://github.com/oppia/oppia/blob/develop/CONTRIBUTING.md#instructions-for-making-a-code-change): Fixed merge conflicts. 

Please review this finally to merged.

seanlip commented on core/controllers/base_test.py:
```diff
@@ -87,6 +87,18 @@ def test_requests_for_invalid_paths(self):
+
+        # Logged out state
+        response = self.testapp.get('/')
+        self.assertEqual(response.status_int, 302)
```
Both these cases should check that the destination location is correct.

526avijitgupta replied:
@seanlip
* For the 1st case (logged out user), checking for `self.testapp.get('/').headers['location']`, contains `gallery`, as it should.

* But for the 2nd case (logged in user), after the `self.login('user@example.com')`, when I check for `self.testapp.get('/').headers['location']`, then it's value is `https://www.google.com/accounts/Logout?continue=http%3A//localhost/` 

I'm not getting what is happening here, since after the login, shouldn't the url location be the one for `my_explorations` ?

seanlip replied:
In cases like this, you will need to debug the flow to get a better sense of what is actually happening. Here is how I did it:

- I tried to raise an exception in RedirectBasedOnLoginStatus, but when I ran the test, the exception did not trigger.
- So I looked at its superclass, base.BaseHandler. I raised an exception in __init__() and it fired. I also noticed that, in this test, the user hasn't really fully logged in (they never chose a username) and so they'll have logged in but not registered, which corresponds to self.partially_logged_in being True.
- I then looked at dispatch() and noted that if the user is partially logged in, we redirect users to the logout URL.

So, the issue here is that the user you're invoking has not fully logged in. Your test will need to register the user first (with a username). Many other backend tests do this already; you can follow the patterns there. I suggest registering the user at the very beginning of the test, and only then testing the two different cases.

Hope this helps!

526avijitgupta replied:
@seanlip :+1:Thank you, for not just finding a solution to this, but also for explaining your debug process.

I'm really getting hold of a lot about the development workflow and the testing process followed at Oppia, thanks to your comments.

Also, yes it worked when I registered the user first.

seanlip replied:
Excellent, I'm happy it's useful! Let me know when you've submitted a new commit, and I'll take another look.

seanlip commented on core/controllers/base_test.py:
```diff
@@ -87,6 +87,18 @@ def test_requests_for_invalid_paths(self):
+        self.login('user@example.com')
+        response = self.testapp.get('/')
+        self.assertEqual(response.status_int, 302)
+
```
FYI, it's good practice while writing this sort of test to do self.logout() at the end (and, more generally, to return things to their original state).

526avijitgupta replied:
Yes. I'm not sure how I missed that.

seanlip commented on main.py:
```diff
@@ -60,6 +61,14 @@ def get(self):
+    "Check whether user is logged in or logged out and perform redirects on '/'"
+    def get(self):
+        if self.user_id and user_services.has_fully_registered(self.user_id):
+            self.redirect('/my_explorations')
```
feconf.MY_EXPLORATIONS_URL, feconf.GALLERY_URL

526avijitgupta replied:
Done.

seanlip commented:
Looking good, thanks! Just a few more small things.

seanlip commented on main.py:
```diff
@@ -60,6 +61,14 @@ def get(self):
         """Handles GET warmup requests."""
         pass
 
+class RedirectBasedOnLoginStatus(base.BaseHandler):
```
Let's call this HomePageRedirectHandler, to keep with the convention of handlers having names that end with 'Handler'.

Also, small nit: can you add an extra newline above, to follow the convention in the code that two newlines separate all top-level classes?

526avijitgupta replied:
@seanlip Yeah I'll do that.

526avijitgupta commented:
@seanlip Updated the changes as you said.

Kindly review :)

seanlip commented:
Actually, noticed a few more things when reviewing:
- the '/' in side_nav.html needs to be changed to '/gallery'.
- the top sentence in admin/admin.html should become "Click here to go to the gallery.", with the link changed accordingly.

Otherwise, it looks good. Thanks, @526avijitgupta!

526avijitgupta commented:
@seanlip Fixed the above mentioned.

Also, thanks for your in-depth review of each commit. I'm glad about not only fixing the issue with your help, but also getting to familiarize myself with the development process and workflow.

seanlip commented:
Looks great, thanks @526avijitgupta! Will merge when Travis checks pass.
